File: bahis/elo.js
/**
 * elo.js
 * -----------------------------------------------------------------
 * Futbol için uyarlanmış ELO rating sistemi (Glicko benzeri
 * güven aralığı eklentisiyle). Dixon-Coles modeliyle paralel
 * çalıştırılıp "iki model aynı fikirde mi" kontrolü için idealdir.
 *
 * Klasik ELO'dan farkı:
 *  - Gol farkına göre K-faktör ağırlıklandırma (büyük skorlar daha
 *    çok puan taşır ama sqrt ile aşırıya kaçmaz - "gol farmalama"
 *    etkisini sınırlar)
 *  - Ev sahibi avantajı sabit offset olarak eklenir
 *  - RD (rating deviation) ile "bu takımın reytingi ne kadar
 *    güvenilir" bilgisi tutulur - az maç oynayan takımlarda RD
 *    yüksek kalır, tahminlere daha az güvenirsin
 * -----------------------------------------------------------------
 */

class EloModel {
  /**
   * @param {Array} matches - [{ home, away, homeGoals, awayGoals, date }]
   *   (Dixon-Coles ile aynı formatta veri kullanabilirsin)
   * @param {Object} opts
   */
  constructor(matches, opts = {}) {
    this.matches = [...matches].sort((a, b) => new Date(a.date) - new Date(b.date));
    this.baseElo = opts.baseElo || 1500;
    this.baseK = opts.baseK || 20;
    this.homeAdv = opts.homeAdv || 65; // ELO puanı cinsinden ev sahibi avantajı
    this.baseRD = opts.baseRD || 350; // yeni takım / az veri belirsizliği
    this.minRD = opts.minRD || 60;

    this.ratings = {}; // { team: { elo, rd, matchCount } }
  }

  _get(team) {
    if (!this.ratings[team]) {
      this.ratings[team] = { elo: this.baseElo, rd: this.baseRD, matchCount: 0 };
    }
    return this.ratings[team];
  }

  // Beklenen kazanma olasılığı (klasik ELO formülü, ev avantajı dahil)
  expectedScore(ratingA, ratingB) {
    return 1 / (1 + Math.pow(10, (ratingB - ratingA) / 400));
  }

  // Gol farkına göre K çarpanı - 1 fark ile 4 fark aynı ağırlıkta olmasın
  // ama sınırsız da büyümesin diye sqrt kullanılıyor (yaygın bir pratik)
  goalDiffMultiplier(goalDiff) {
    return Math.sqrt(Math.max(goalDiff, 1));
  }

  fit() {
    for (const m of this.matches) {
      const home = this._get(m.home);
      const away = this._get(m.away);

      // Maç sonucu: 1 = ev kazandı, 0.5 = berabere, 0 = deplasman kazandı
      let actualHome;
      if (m.homeGoals > m.awayGoals) actualHome = 1;
      else if (m.homeGoals === m.awayGoals) actualHome = 0.5;
      else actualHome = 0;

      const expectedHome = this.expectedScore(
        home.elo + this.homeAdv,
        away.elo
      );

      const goalDiff = Math.abs(m.homeGoals - m.awayGoals);
      const mult = this.goalDiffMultiplier(goalDiff);

      // RD arttıkça (belirsizlik yüksekken) K-faktör büyür - yeni takımlar
      // gerçek seviyesine daha hızlı yaklaşır
      const kHome = this.baseK * mult * (home.rd / this.baseRD);
      const kAway = this.baseK * mult * (away.rd / this.baseRD);

      home.elo += kHome * (actualHome - expectedHome);
      away.elo += kAway * ((1 - actualHome) - (1 - expectedHome));

      // RD zamanla daralır (daha çok maç = daha güvenilir reyting)
      home.matchCount++;
      away.matchCount++;
      home.rd = Math.max(this.minRD, this.baseRD / Math.sqrt(home.matchCount));
      away.rd = Math.max(this.minRD, this.baseRD / Math.sqrt(away.matchCount));
    }
    return this.ratings;
  }

  // Maç öncesi olasılık tahmini (1X2)
  predictMatch(home, away) {
    const h = this._get(home);
    const a = this._get(away);

    const pHomeWin = this.expectedScore(h.elo + this.homeAdv, a.elo);
    // Beraberlik olasılığını basit bir yaklaşımla ELO farkından türetiyoruz:
    // fark küçükse beraberlik ihtimali yüksek, fark büyüdükçe düşer
    const eloDiff = Math.abs(h.elo + this.homeAdv - a.elo);
    const pDraw = Math.max(0.18, 0.32 - eloDiff / 1000);

    // pHomeWin aslında "ev kazanır ya da berabere" tipi bir skor değil,
    // düzeltip üç olasılığı normalize ediyoruz
    let pHome = pHomeWin - pDraw / 2;
    let pAway = (1 - pHomeWin) - pDraw / 2;
    pHome = Math.max(pHome, 0.02);
    pAway = Math.max(pAway, 0.02);

    const total = pHome + pDraw + pAway;
    return {
      "1": pHome / total,
      X: pDraw / total,
      "2": pAway / total,
      confidence: {
        home: { elo: h.elo, rd: h.rd, matches: h.matchCount },
        away: { elo: a.elo, rd: a.rd, matches: a.matchCount },
        // rd toplamı yüksekse (az veri) bu tahmine az güven
        reliable: h.rd < 150 && a.rd < 150,
      },
    };
  }

  // Dixon-Coles çıktısıyla karşılaştırıp uyum kontrolü yapar
  static agreementCheck(eloProbs, dixonColesProbs, threshold = 0.15) {
    const diffs = {
      "1": Math.abs(eloProbs["1"] - dixonColesProbs["1"]),
      X: Math.abs(eloProbs.X - dixonColesProbs.X),
      "2": Math.abs(eloProbs["2"] - dixonColesProbs["2"]),
    };
    const maxDiff = Math.max(diffs["1"], diffs.X, diffs["2"]);
    return {
      diffs,
      maxDiff,
      // iki model de benzer sonuca varıyorsa güven yüksek
      agree: maxDiff <= threshold,
    };
  }

  ranking() {
    return Object.entries(this.ratings)
      .sort((a, b) => b[1].elo - a[1].elo)
      .map(([team, r]) => ({ team, elo: Math.round(r.elo), rd: Math.round(r.rd), matches: r.matchCount }));
  }
}

module.exports = { EloModel };

// ---------- Kullanım örneği ----------
if (require.main === module) {
  const sampleMatches = [
    { home: "Galatasaray", away: "Fenerbahçe", homeGoals: 2, awayGoals: 1, date: "2026-03-01" },
    { home: "Fenerbahçe", away: "Beşiktaş", homeGoals: 1, awayGoals: 1, date: "2026-02-20" },
    { home: "Beşiktaş", away: "Trabzonspor", homeGoals: 0, awayGoals: 2, date: "2026-02-10" },
    { home: "Trabzonspor", away: "Galatasaray", homeGoals: 1, awayGoals: 3, date: "2026-01-25" },
    { home: "Galatasaray", away: "Beşiktaş", homeGoals: 3, awayGoals: 0, date: "2026-01-10" },
    { home: "Fenerbahçe", away: "Trabzonspor", homeGoals: 2, awayGoals: 0, date: "2025-12-15" },
  ];

  const elo = new EloModel(sampleMatches);
  elo.fit();

  console.log("Sıralama:");
  console.table(elo.ranking());

  const pred = elo.predictMatch("Galatasaray", "Fenerbahçe");
  console.log("\nGalatasaray - Fenerbahçe (ELO):", pred);

  // Dixon-Coles çıktısıyla karşılaştırma örneği (dixonColes.js'den gelen sonuç)
  const dcProbs = { "1": 0.714, X: 0.173, "2": 0.113 };
  const check = EloModel.agreementCheck(
    { "1": pred["1"], X: pred.X, "2": pred["2"] },
    dcProbs
  );
  console.log("\nİki model uyum kontrolü:", check);
}

File: bahis/elo.py
"""ELO rating — `elo.js` ile aynı aritmetik. Emir yok."""
from __future__ import annotations

import math
from datetime import datetime, timedelta
from functools import lru_cache
from zoneinfo import ZoneInfo

from bahis.dixon_coles import (
    UPCOMING_DAYS,
    UPCOMING_LIMIT,
    _pct,
    _round_p,
    _train_rows,
    _when,
)
from bahis.league import all_matches, team_info

TR = ZoneInfo("Europe/Istanbul")
BASE_ELO = 1500.0
BASE_K = 20.0
HOME_ADV = 65.0
BASE_RD = 350.0
MIN_RD = 60.0


class EloModel:
    def __init__(self, matches: list[dict], opts: dict | None = None):
        opts = opts or {}
        self.matches = sorted(matches, key=lambda m: m["date"])
        self.base_elo = opts.get("baseElo", BASE_ELO)
        self.base_k = opts.get("baseK", BASE_K)
        self.home_adv = opts.get("homeAdv", HOME_ADV)
        self.base_rd = opts.get("baseRD", BASE_RD)
        self.min_rd = opts.get("minRD", MIN_RD)
        self.ratings: dict[str, dict] = {}

    def _get(self, team: str) -> dict:
        if team not in self.ratings:
            self.ratings[team] = {
                "elo": self.base_elo,
                "rd": self.base_rd,
                "matchCount": 0,
            }
        return self.ratings[team]

    @staticmethod
    def expected_score(rating_a: float, rating_b: float) -> float:
        return 1 / (1 + 10 ** ((rating_b - rating_a) / 400))

    @staticmethod
    def goal_diff_multiplier(goal_diff: float) -> float:
        return math.sqrt(max(goal_diff, 1))

    def apply_match(self, m: dict) -> None:
        """Tek maçı işler — walk-forward: önce predict, sonra apply."""
        home = self._get(m["home"])
        away = self._get(m["away"])
        if m["homeGoals"] > m["awayGoals"]:
            actual_home = 1.0
        elif m["homeGoals"] == m["awayGoals"]:
            actual_home = 0.5
        else:
            actual_home = 0.0
        expected_home = self.expected_score(home["elo"] + self.home_adv, away["elo"])
        goal_diff = abs(m["homeGoals"] - m["awayGoals"])
        mult = self.goal_diff_multiplier(goal_diff)
        k_home = self.base_k * mult * (home["rd"] / self.base_rd)
        k_away = self.base_k * mult * (away["rd"] / self.base_rd)
        home["elo"] += k_home * (actual_home - expected_home)
        away["elo"] += k_away * ((1 - actual_home) - (1 - expected_home))
        home["matchCount"] += 1
        away["matchCount"] += 1
        home["rd"] = max(self.min_rd, self.base_rd / math.sqrt(home["matchCount"]))
        away["rd"] = max(self.min_rd, self.base_rd / math.sqrt(away["matchCount"]))

    def fit(self) -> dict:
        for m in self.matches:
            self.apply_match(m)
        return self.ratings

    def predict_match(self, home: str, away: str) -> dict:
        h, a = self._get(home), self._get(away)
        p_home_win = self.expected_score(h["elo"] + self.home_adv, a["elo"])
        elo_diff = abs(h["elo"] + self.home_adv - a["elo"])
        p_draw = max(0.18, 0.32 - elo_diff / 1000)
        p_home = max(p_home_win - p_draw / 2, 0.02)
        p_away = max((1 - p_home_win) - p_draw / 2, 0.02)
        total = p_home + p_draw + p_away
        return {
            "1": p_home / total,
            "X": p_draw / total,
            "2": p_away / total,
            "confidence": {
                "home": {"elo": h["elo"], "rd": h["rd"], "matches": h["matchCount"]},
                "away": {"elo": a["elo"], "rd": a["rd"], "matches": a["matchCount"]},
                "reliable": h["rd"] < 150 and a["rd"] < 150,
            },
        }

    @staticmethod
    def agreement_check(elo_probs: dict, dc_probs: dict, threshold: float = 0.15) -> dict:
        diffs = {
            "1": abs(elo_probs["1"] - dc_probs["1"]),
            "X": abs(elo_probs["X"] - dc_probs["X"]),
            "2": abs(elo_probs["2"] - dc_probs["2"]),
        }
        max_diff = max(diffs.values())
        return {"diffs": diffs, "maxDiff": max_diff, "agree": max_diff <= threshold}

    def ranking(self) -> list[dict]:
        return [
            {
                "team": t,
                "elo": round(r["elo"]),
                "rd": round(r["rd"]),
                "matches": r["matchCount"],
            }
            for t, r in sorted(self.ratings.items(), key=lambda kv: kv[1]["elo"], reverse=True)
        ]


@lru_cache(maxsize=8)
def _fitted_for(league: str) -> EloModel:
    from bahis.leagues_cfg import set_league
    set_league(league)
    model = EloModel(_train_rows())
    model.fit()
    return model


def _fitted() -> EloModel:
    from bahis.leagues_cfg import current_league
    return _fitted_for(current_league())


def _card(m: dict, pred: dict) -> dict:
    hi, ai = m["home"], m["away"]
    mr = {k: _round_p(pred[k]) for k in ("1", "X", "2")}
    pick = max(mr, key=mr.get)
    if pick == "1":
        text = f"{hi['name']} kazanır"
    elif pick == "2":
        text = f"{ai['name']} kazanır"
    else:
        text = "Beraberlik"
    conf = pred["confidence"]
    return {
        "id": m["id"],
        "week": m.get("week"),
        "kickoff": m.get("kickoff"),
        "when": _when(m.get("kickoff")),
        "venue": m.get("venue"),
        "home": hi,
        "away": ai,
        "odds": m.get("odds") or {},
        "matchResult": mr,
        "doubleChance": {
            "1X": _round_p(mr["1"] + mr["X"]),
            "12": _round_p(mr["1"] + mr["2"]),
            "X2": _round_p(mr["X"] + mr["2"]),
        },
        "pick": pick,
        "text": text,
        "pct": _pct(mr[pick]),
        "elo": {
            "home": round(conf["home"]["elo"]),
            "away": round(conf["away"]["elo"]),
            "rd_h": round(conf["home"]["rd"]),
            "rd_a": round(conf["away"]["rd"]),
            "n_h": conf["home"]["matches"],
            "n_a": conf["away"]["matches"],
            "reliable": bool(conf["reliable"]),
        },
        "value": [],
    }


def upcoming_preds(team: str | None = None, limit: int = UPCOMING_LIMIT) -> dict:
    now = datetime.now(TR)
    until = now + timedelta(days=UPCOMING_DAYS)
    now_s, until_s = now.isoformat(), until.isoformat()
    model = _fitted()
    rows = []
    for m in all_matches():
        if m.get("played"):
            continue
        ko = m.get("kickoff") or ""
        if ko < now_s[:10] or ko > until_s:
            continue
        if team and team not in (m["home"]["key"], m["away"]["key"]):
            continue
        pred = model.predict_match(m["home"]["key"], m["away"]["key"])
        rows.append(_card(m, pred))
        if len(rows) >= limit:
            break
    used = {r["home"]["key"] for r in rows} | {r["away"]["key"] for r in rows}
    strengths = []
    for r in model.ranking():
        if r["team"] not in used:
            continue
        info = team_info(r["team"])
        strengths.append({
            "key": r["team"],
            **info,
            "attack": r["elo"],
            "defense": r["rd"],
        })
    return {
        "ok": True,
        "model": "elo",
        "note": f"{len(model.matches)} maç · ev +{int(HOME_ADV)} ELO · RD güven",
        "matches_used": len(model.matches),
        "horizon_days": UPCOMING_DAYS,
        "updated": now.isoformat(timespec="seconds"),
        "n": len(rows),
        "preds": rows,
        "strengths": strengths,
    }
